Refactor the Resolver + CLI

AntonC9018 opened the pull request:
Don't merge this yet, I'm still refactoring the code.

**Overview:**

- `orphanArgument` token type, which means a positional argument after any named arguments.
- Tests for the binder module.
- Implemented groups for commands. 
- Renamed and refactored some stuff.

**More on the groups:**

How it works is if there are any extra positional arguments, that argument will be treated as a nested command name. It will try to match any command that references the "command group" via a pointer field of that command context type. Before that nested command is invoked, the intermediate command is also executed.

The command tree or the command graph is created at compile time and is not available to traditional reflection at runtime. This can be added if needed tho. You'd just take all of the info on that graph (it's in an anonymous CTFE lambda) and just extract it in a template or something.

Currently, there is no way for the intermediate command to tell whether it's getting invoked as an intermediate command or as a terminal command, but I suppose we could just call a different method if it's not terminal.

Arguments are matched in this order:
- Until the minimum required number of positional arguments is reached, the arguments are fed into the command parser and are bound to the argument fields.
- Even if there still are positional optional arguments, we first try to match the next nested command from the list of the child commands of the current command type, then we try to bind the positional argument.
- If no positional arguments are possible, the arguments must be matched to a command, otherwise an error is produced.

Any of the help arguments `-help --help -h --h /? /h /help` are treated in a special way and currently result in a fix-me message being printed.

AntonC9018 commented:
@BradleyChatha can you disable the github actions so it doesn't work in vain?

BradleyChatha commented:
Actions are disabled for now

AntonC9018 commented:
I think I'm done with this module for now. I've refactored it into a state machine.
The code is very metaprogramming-heavy, though I'm sure it can be simplified.

The type graph thing is probably too compilcated, needlessly. Even though the sole purpose of mapping types to indices at compile time is so that the execution time of the algorithm is linear (it's quadratic if you'd just loop through the AliasSeq and find the type there), it can only matter if the number of types is significant. For a small number of types, a dumb version will be just as fast.

Another idea of how to map indices to types is via template resolutions, but this would rely on whether that is linear in the first place, which I don't think it is.

When immutable AA fields in templates become possible in D, that code can be simplified to just use those for the given mapping instead of restricting that info to CTFE or possibly converting it into a template name lookup hack via a mixin, if you happen to need those mappings in other places at compile time (commented-out code in `groups.d`, because we happen not to need it).

Things left to implement:

- Partial matches, pretty easy stuff. Just hook into the function that tries to match names (there is one for root names and another one for the subcommand names). Just modify that logic as you see fit. The only problematic bit is that you can't really uninline it, unless you do a mixin. There is a comment about this, but basically, you hit some pretty annoying frame issues.
- Command help, really easy, all of the info is there. You get the command type index and a notification that the help has been requested (see "special things"), in its simplest form, just print the help of that command once you're there.
- Removing false information from the readme, putting down the correct information, and also fixing the examples again.

I think it will be worth it to try and work on the API a little bit. In particular, using all the bits and pieces individually, starting from writing a small simple app and scaling up to an infrastructure. I think we should make sure the lib is easy to use at any scale.

BradleyChatha commented on subprojects/introspect/source/jcli/introspect/data.d:
```diff
@@ -538,11 +537,19 @@ import std.meta;
-ArgFlags foldArgumentFlags(udas...)()
+// Another NOTE:
+// We still need the static here?? wtf?
+static ArgFlags foldArgumentFlags(udas...)()
```
Might be falling victim to this amazingly annoying feature of templates: https://dlang.org/spec/template.html#implicit-nesting

BradleyChatha commented on subprojects/introspect/source/jcli/introspect/groups.d:
```diff
@@ -0,0 +1,347 @@
+}
+
+template TypeGraph(Types...)
+{
```
Might be a waste of time, idk, but I wonder if we should make a function to output dot files for [graphviz](https://graphviz.org/) so we can visualise the TypeGraph.

AntonC9018 replied:
You can do that if you want, I'm not the one who's going to do it

BradleyChatha reviewed (commented):
Looking good so far.

Had a decent look over the code, and once I'm done with work tomorrow I'll try to properly dive into it to make sure I understand everything that's going on, and then get it all merged if I see nothing major or that I want addressing.

And then I'll also write up those things for the help text :)

BradleyChatha commented:
Also I'll try to address any questions you've left in the code as well

AntonC9018 commented on source/jcli/cli.d:
```diff
@@ -2,327 +2,1675 @@ module jcli.cli;
+    struct StorageItem
+    {
+        int typeIndex;
+        void* commandPointer;
```
It's stored internally this way, in fact. We could and should expose it to the user via a type-safe wrapper. Using it directly like this internally is totally ok imo.

AntonC9018 replied:
The index can be mapped back to the type by just indexing Types. The reason I store it as an index is because the graph works with indices+it's context independent, essentally

AntonC9018 replied:
What I mean by that, is that the context struct is just data. Whatever is on top of it decides what to do with in on its own.

AntonC9018 commented:
@BradleyChatha ping me here or on discord any time if you have any questions to me too

BradleyChatha commented:
Will do :)

AntonC9018 commented:
I've cleaned up the code a little bit, and somewhat started on the type context idea. The internal hassle currently is that there are 4 different contexts to manage, which can get a bit confusing.

I've also just added an example where the commands are collected from the given module. The API level is at the easy to use and pretty complete level, even though internally, it is somewhat messy and still needs refactoring imo.

The help messages suck at the moment too, I hope you can get them sorted soon.

BradleyChatha commented:
Just to be transparent, I'll probably take a look at the weekend now. I've had no energy to do anything this past week so I'm completely drained by the end of the workday T.T

Apologies, and I'll try to get this through as soon as I'm able to manage.

BradleyChatha commented on new_examples/across_modules/source/commands.d:
```diff
@@ -0,0 +1,94 @@
+    // Examples:
+    // @("Does stuff")
+    // @Rename("debug")
+    // @Hidden
```
I might do this as part of the help text improvements

BradleyChatha commented on source/jcli/cli.d:
```diff
@@ -2,327 +2,1929 @@ module jcli.cli;
+    }
+    catch (Exception exc)
+    {
+        return typeof(return)(0, exc);
```
Return 0 on error?

AntonC9018 replied:
It returns the exception, the user should check that first. I didn't wrap it in a getter, because so far it's only been used internally and extra fluff for data objects is just annoying to me. You can do a SumType here too to ne able to match over the result

BradleyChatha commented on source/jcli/cli.d:
```diff
@@ -2,327 +2,1929 @@ module jcli.cli;
+        }
+        else
+        {
+            return typeof(return)(0, null);
```
Should this branch become a compile-time error, instead of silently failing to call onExecute?

My main concern is someone managing to mess up their definition of onExecute and getting no feedback from JCLI one what went wrong.

AntonC9018 replied:
The user may want to hook their command into another command, such that the next command depends on the logic of the intermediate command. So I think it's ok, given we go bottom-up (`ParentCommand`) instead of top-down (manually listing children). I think this should only be an error when the children were listed explicitly, which I have not implemented.

AntonC9018 replied:
@BradleyChatha sorry, misread that.
Noe that I have added a separate callback for when it's executed as an intermediate command, we can indeed make that an error, or default to printing the help of children, if there are any.

BradleyChatha commented on subprojects/introspect/source/jcli/introspect/data.d:
```diff
@@ -130,11 +113,88 @@ template getArgumentFieldSymbol(TCommand, alias argumentInfoOrFieldPath)
+// template, whether it should be a getter, and enum, immutable, or an alias.
+// It feels to me like it's only philosophical at this point.
+// So I'm just going to do a single thing and call it a day.
+// Whatever I do feels wrong tho.
```
A template is fine. Going with a `struct` makes it a bit easier to populate from CTFE functions though, so I'd say it depends on your needs.

BradleyChatha commented on subprojects/introspect/source/jcli/introspect/groups.d:
```diff
@@ -0,0 +1,332 @@
+        }
+
+        // Extra info, not actually used
+        static if (0)
```
Is this for debugging purposes? Otherwise it might be wise to consider if we should just remove it instead of disable it.

BradleyChatha replied:
We'll keep it for now then, and maybe remove it once we're certain the code is in a more stable state (i.e. we won't be working on it much more).

BradleyChatha commented:
> I think it will be worth it to try and work on the API a little bit. In particular, using all the bits and pieces individually, starting from writing a small simple app and scaling up to an infrastructure. I think we should make sure the lib is easy to use at any scale.

Need ideas to implement. Maybe something simple like reimplementing a lot of common command line applications, but in JCLI, to show it off a little.

One project I wanted to do with the original JCLI was to reimplement Dub's CLI, and see what pain points and blocked popped up, but never got around to it.

AntonC9018 commented on subprojects/introspect/source/jcli/introspect/data.d:
```diff
@@ -130,11 +113,88 @@ template getArgumentFieldSymbol(TCommand, alias argumentInfoOrFieldPath)
+        static if (is(typeof(rawCommandUDA) : string))
+        {
+            enum flags = CommandFlags.stringAttribute | CommandFlags.givenValue;
+            enum string udaValue = rawCommandUDA;
```
Aren't strings already immutable? I'm pretty sure it doesn't apply to them.

BradleyChatha replied:
Just checked the ASM output, and yes you're right! (Surprisingly, DMD also doesn't duplicate it!)

AntonC9018 commented on source/jcli/cli.d:
```diff
@@ -2,327 +2,1929 @@ module jcli.cli;
+            }
+
+            // We don't bother with subcommands if the current command has no children.
+            // Is this a hack? I'm not sure.
```
@BradleyChatha I was referring more or less to the fact that I used the same function for both a single command execution and command discovery within modules, matching, and execution. But the fact that this piece of logic is hit even for the multiple-command case, I think justifies that it's alright.

I was debating whether faking that a root command has been matched in the single-command use case is a hack or not, but then decided it's ok precisely because that piece of logic is hit for both.

BradleyChatha commented:
> @BradleyChatha I have done an example of simple usage and an example of module-based usage, since I left that comment. It's just that I think that if we claim our thing is modular and easy to use, we should prove it really is by trying to use it as such, which way we'll discover flaws and be able to adjust or refactor it accordingly.
> 
> Not necessarily big projects, demos like in your examples work, it's just that this should be done for each module that you think can be used stand-alone. Just do such demos and you'll discover things that feel out of place, something or extra fluff that the user ideally does not need to worry about.
> 
> Example: the old command parser (I think) had the extra requirement of a command UDA being there on the command to be able to use it. I think this is what's called a quirk because the user does not care about any of the info that may be provided via that attribute to the application, but it still needed to be present, which was annoying.
> 
> Another example: the old CLI only worked with modules, taking on command discovery, which feels like a burden when all you want is to execute a single command.

Yeah, it's just a good idea in general to 'practice what we preach'. The more examples we can come up with the better.

> Why do you need the test runner for those examples btw? Integration tests? Or just making sure it compiles? Because tests from the source code are plenty. I think demos are fine to stay as demos.

It was mostly just to make sure that they actually compile. Too many times have I seen projects with broken examples. I didn't want to end up as one of them :D 

> Even then, I think handcrafted demos should be separate from actual rigorous tests.

Perhaps we can make the distinction between examples of module usage, and demos of a fully-fledged program. The former can be handled with unittests, the latter with a test runner.
